### PR TITLE
tensorflow: update the userguide with the latest release details

### PR DESCRIPTION
**Pull request by snadampal**
TensorFlow deep learning container, wheel, and docker image details are updated. xla section is removed because the latest release integrated openXLA which requires changes in the build and usage. will update the userguide with XLA details once openXLA is validated on Graviton.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by provanshu** [comment]
Thank you for looking into this. The clang and gcc builds work great!

File: machinelearning/tensorflow.md [hunk 1]
```diff
@@ -265,16 +272,93 @@ export TF_NEED_OPENCL_SYCL=0
...
+# Bazel version required depends on the TensorFlow version, please install the correct one
+# https://github.com/tensorflow/tensorflow/blob/master/.bazelversion captures the bazel version details
+# For example, tensorflow 2.16.1 build requires bazel 6.5.0
+cd $HOME
+mkdir bazel
+cd bazel
+wget https://github.com/bazelbuild/bazel/releases/download/<bazelversion>/bazel-<bazelversion>-linux-arm64
+mv bazel-<bazelversion>-linux-arm64 bazel
+chmod a+x bazel
+export PATH=/home/ubuntu/bazel/:$PATH
```
Comment (CristianPrundeanu) [inline machinelearning/tensorflow.md:318]: Minor: better to install bazel into the standard /usr/local/bin so you don't need to modify PATH every time
```
BAZEL=6.5.0  # modify version as needed
wget https://github.com/bazelbuild/bazel/releases/download/$BAZEL/bazel-$BAZEL-linux-arm64 -O /usr/local/bin/bazel
chmod +x /usr/local/bin/bazel
```

File: machinelearning/tensorflow.md [hunk 2]
```diff
@@ -30,14 +30,14 @@ docker run -p 8501:8501 --name tfserving_resnet --mount type=bind,source=/tmp/re
 pip install tensorflow
 ```
 
-TensorFlow wheel supports the default eigen backend for Graviton CPUs, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
+Starting from version 2.14.0, TensorFlow wheel enabled onednn+acl backend by default on Graviton CPUs which typically provides better performance than eigen backend. For previous versions of TensorFlow, onednn+acl can be enabled by setting the below TF environment variable
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:33]: I would reword this sentence for better reading.
"On Graviton CPUs, onednn+acl is the default backend in TensorFlow v2.14.0 wheel. This backend typically provides better performance than eigen backend. For prev..."

File: machinelearning/tensorflow.md [hunk 3]
```diff
@@ -30,14 +30,14 @@ docker run -p 8501:8501 --name tfserving_resnet --mount type=bind,source=/tmp/re
...
 ```
 export TF_ENABLE_ONEDNN_OPTS=1
 ```
 
 **Using Docker hub container**
 
-As of June 2023, Docker hub images from armswdev are based on TensorFlow 2.12.0, but also include additional downstream optimizations and experimental features. These are avaiable for trying out the experimental downstream features and provide early feedback.
+As of May 2024, Docker hub images from armswdev are based on TensorFlow 2.15.1, but also include additional downstream optimizations and experimental features. These are avaiable for trying out the experimental downstream features and provide early feedback.
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:40]: redundant, can delete:
"These are avaiable for trying out the experimental downstream features and provide early feedback."

File: machinelearning/tensorflow.md [hunk 4]
```diff
@@ -55,7 +55,7 @@ It is highly recommended to use the AMIs based on Linux Kernel 5.10 and beyond f
 
 AWS DLCs come with all the optimizations enabled, so, there are no additional runtime configurations required. Where as for the python wheels and the docker hub images, enable the below runtime configurations to achieve the best performance.
 ```
-# The default runtime backend for tensorflow is Eigen, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
+# For TensorFlow versions older than 2.14.0, the default runtime backend is Eigen, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:58]: This was mentioned above too. Can delete?

File: machinelearning/tensorflow.md [hunk 5]
```diff
@@ -291,10 +375,12 @@ sudo apt install maven default-jdk
 # Install bazel for aarch64
 # Bazel version required depends on the TensorFlow version, please install the correct one
 # https://github.com/tensorflow/tensorflow/blob/master/.bazelversion captures the bazel version details
+# For example, tensorflow 2.16.1 build requires bazel 6.5.0
+cd $HOME
 mkdir bazel
 cd bazel
-wget https://github.com/bazelbuild/bazel/releases/download/5.1.1/bazel-5.1.1-linux-arm64
-mv bazel-5.1.1-linux-arm64 bazel
+wget https://github.com/bazelbuild/bazel/releases/download/<bazelversion>/bazel-<bazelversion>-linux-arm64
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:382]: Can you add a step to define `bazelversion`?

File: machinelearning/tensorflow.md [hunk 6]
```diff
@@ -217,21 +216,29 @@ OPENMP DISPLAY ENVIRONMENT END
...
 # Install bazel for aarch64
+# Bazel version required depends on the TensorFlow version, please install the correct one
+# https://github.com/tensorflow/tensorflow/blob/master/.bazelversion captures the bazel version details
+# For example, tensorflow 2.16.1 build requires bazel 6.5.0
+cd $HOME
 mkdir bazel
 cd bazel
-wget https://github.com/bazelbuild/bazel/releases/download/5.1.1/bazel-5.1.1-linux-arm64
-mv bazel-5.1.1-linux-arm64 bazel
+wget https://github.com/bazelbuild/bazel/releases/download/<bazelversion>/bazel-<bazelversion>-linux-arm64
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:240]: Can you add a step to define `bazelversion`?

File: machinelearning/tensorflow.md [hunk 7]
```diff
@@ -246,7 +253,7 @@ export HOST_C_COMPILER=(which gcc)
 export HOST_CXX_COMPILER=(which g++)
 export PYTHON_BIN_PATH=(which python)
 export USE_DEFAULT_PYTHON_LIB_PATH=1
-export TF_ENABLE_XLA=1
+export TF_ENABLE_XLA=0
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:256]: is it required? shoud be set to `0`?

File: machinelearning/tensorflow.md [hunk 8]
```diff
@@ -265,16 +272,93 @@ export TF_NEED_OPENCL_SYCL=0
...
+# so, enter "n" for Clang build option.
+# Do you want to use Clang to build TensorFlow? [Y/n]: n
 ./configure
 
 # Issue bazel build command with 'mkl_aarch64' config to enable onednn+acl backend
 bazel build --verbose_failures -s --config=mkl_aarch64  //tensorflow/tools/pip_package:build_pip_package //tensorflow:libtensorflow_cc.so //tensorflow:install_headers
 
 # Create and install the wheel
-./bazel-bin/tensorflow/tools/pip_package/build_pip_package ./wheel-TF2.12.0-py3.8-aarch64
+./bazel-bin/tensorflow/tools/pip_package/build_pip_package ./wheel-TF<version>-aarch64
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:285]: can you add a step to define `version`?

File: machinelearning/tensorflow.md [hunk 9]
```diff
@@ -265,16 +272,93 @@ export TF_NEED_OPENCL_SYCL=0
...
+```
+# Install the llvm version recommended in the TensorFlow release
+# llvm-17.0.2 is recommended for TF v2.16
+cd $HOME
+mkdir llvm_install
+cd llvm_install
+wget https://github.com/llvm/llvm-project/releases/download/llvmorg-17.0.2/clang+llvm-17.0.2-aarch64-linux-gnu.tar.xz .
+tar -xvf clang+llvm-17.0.2-aarch64-linux-gnu.tar.xz clang+llvm-17.0.2-aarch64-linux-gnu/
+sudo cp -r clang+llvm-17.0.2-aarch64-linux-gnu/* /usr
+clang --version
```
Comment (RamaMalladiAWS) [inline machinelearning/tensorflow.md:302]: The below steps are repeat of what we have under `gcc`. Possibly you can combine?

**Review by RamaMalladiAWS — CHANGES_REQUESTED** [review]
Looks good, some changes I have suggested.

**Comment by snadampal** [comment]
@CristianPrundeanu , @RamaMalladiAWS , thanks for the review and suggestion. I have updated the PR.

**Review by RamaMalladiAWS — APPROVED** [review]
